### PR TITLE
Merge 2.6.6 source code into 2.7

### PR DESCRIPTION
**Pull request by mercyblitz**
Merge 2.6.6 source code into 2.7

### CODE REVIEW

**Comment by chickenlj** [comment]
- spring-context-support去除，直接把源码加进来了
- 2.7中增加了ApplicationConfig只有一个的限制，相应的改动，主要是涉及UT
- org.apache.dubbo相关的包名改动
- https://github.com/apache/incubator-dubbo/issues/1641
- https://github.com/apache/incubator-dubbo/issues/2235
- https://github.com/apache/incubator-dubbo/issues/1654

- https://github.com/apache/incubator-dubbo/milestone/21?closed=1
- https://github.com/apache/incubator-dubbo-spring-boot-project/issues/319

**Comment by chickenlj** [comment]
This patch is a little too big, maybe the next time we can split them into small PRs. 
One PR solve one issue.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/AbstractAnnotationConfigBeanBuilder.java [hunk 1]
```diff
@@ -16,14 +16,13 @@
  */
 package org.apache.dubbo.config.spring.beans.factory.annotation;
 
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
 import org.apache.dubbo.config.AbstractInterfaceConfig;
 import org.apache.dubbo.config.ApplicationConfig;
 import org.apache.dubbo.config.ModuleConfig;
 import org.apache.dubbo.config.MonitorConfig;
 import org.apache.dubbo.config.RegistryConfig;
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/AbstractAnnotationConfigBeanBuilder.java]: @mercyblitz  should you import the codestyle/dubbo_codestyle_for_idea.xml into idea codestyle.

The order of  org.apache.dubbo is firstly.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/DubboConfigBindingBeanPostProcessor.java [hunk 2]
```diff
@@ -16,15 +16,14 @@
  */
 package org.apache.dubbo.config.spring.beans.factory.annotation;
 
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/DubboConfigBindingBeanPostProcessor.java]: As above metioned, the order should be changed

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java [hunk 3]
```diff
@@ -17,44 +17,69 @@
...
  */
-public class DubboConfigConfigurationSelector implements ImportSelector, Ordered {
+public class DubboConfigConfigurationRegistrar implements ImportBeanDefinitionRegistrar, EnvironmentAware, Ordered {
+
+    /**
+     * The property name of {@link EnableDubboConfig}'s {@link Ordered order}
+     */
+    public static final String DUBBO_CONFIG_ORDER_PROPERTY_NAME = "dubbo.config.order";
+
+    private int order;
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java]: questions:
1. I am not sure If we need setOrder to set.
2. is it need a default value?

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java [hunk 4]
```diff
@@ -17,44 +17,69 @@
...
                 importingClassMetadata.getAnnotationAttributes(EnableDubboConfig.class.getName()));
 
         boolean multiple = attributes.getBoolean("multiple");
 
         if (multiple) {
-            return of(DubboConfigConfiguration.Multiple.class.getName());
+            registerBeans(registry, DubboConfigConfiguration.Multiple.class);
         } else {
-            return of(DubboConfigConfiguration.Single.class.getName());
+            registerBeans(registry, DubboConfigConfiguration.Single.class);
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java]: 这种方式和原来ImportSelector方式比，除了简洁之外，有带来其他好处吗？

Compared with the original ImportSelector method, does this method bring other benefits besides being concise?

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/AbstractAnnotationConfigBeanBuilder.java (hunk 1 above)
Comment (mercyblitz) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/AbstractAnnotationConfigBeanBuilder.java]: My IDEA is using the `codestyle/dubbo_codestyle_for_idea.xml` as a code style template.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java (hunk 3 above)
Comment (cvictory) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java]: I have seen the default value in setEnvironment method.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java [hunk 5]
```diff
@@ -17,44 +17,69 @@
...
 
     @Override
     public int getOrder() {
-        return HIGHEST_PRECEDENCE;
+        return order;
     }
 
+    @Override
+    public void setEnvironment(Environment environment) {
+        this.order = environment.getProperty(DUBBO_CONFIG_ORDER_PROPERTY_NAME, int.class, LOWEST_PRECEDENCE);
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboConfigConfigurationRegistrar.java]: Why introduce this setting ？ Is there a issue or give us some context to this.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/properties/DefaultDubboConfigBinder.java [hunk 6]
```diff
@@ -37,11 +36,12 @@
         dataBinder.setIgnoreInvalidFields(isIgnoreInvalidFields());
         dataBinder.setIgnoreUnknownFields(isIgnoreUnknownFields());
         // Get properties under specified prefix from PropertySources
-        Map<String, String> properties = getSubProperties(getPropertySources(), prefix);
+        Map<String, Object> properties = getSubProperties(getPropertySources(), prefix);
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/properties/DefaultDubboConfigBinder.java]: I find the code has no diff

Comment (cvictory) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/properties/DefaultDubboConfigBinder.java]: why?

Comment (cvictory) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/properties/DefaultDubboConfigBinder.java]: Ignore this.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/util/AnnotationUtils.java [hunk 7]
```diff
@@ -39,13 +53,186 @@
...
+     * @return If present , return <code>true</code> , or <code>false</code>
+     * @since 2.6.6
+     */
+    public static <A extends Annotation> boolean isPresent(Method method, Class<A> annotationClass) {
+
+        Map<ElementType, List<A>> annotationsMap = findAnnotations(method, annotationClass);
+
+        return !annotationsMap.isEmpty();
+
+    }
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/util/AnnotationUtils.java]: I found is is no usage. There is no class to invoke this method.

Comment (cvictory) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/util/AnnotationUtils.java]: ignore .  there are some code that is introduced by org.springframework.core.annotation.AnnotationUtils.

File: dubbo-config/dubbo-config-spring/src/test/java/org/apache/dubbo/config/spring/beans/factory/annotation/AnnotationPropertyValuesAdapterTest.java [hunk 8]
```diff
@@ -142,7 +142,7 @@ public String convert(String[] source) {
     private static class TestBean {
 
         @Reference(
-                interfaceClass = DemoService.class, interfaceName = "org.apache.dubbo.config.spring.api.DemoService", version = "${version}", group = "group",
+                interfaceClass = DemoService.class, interfaceName = "com.alibaba.dubbo.config.spring.api.DemoService", version = "${version}", group = "group",
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/test/java/org/apache/dubbo/config/spring/beans/factory/annotation/AnnotationPropertyValuesAdapterTest.java]: Why change to com.alibaba but not org.apache?

File: dubbo-config/dubbo-config-spring/src/test/java/org/apache/dubbo/config/spring/beans/factory/config/YamlPropertySourceFactory.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.Properties;
+import java.util.regex.Pattern;
+
+/**
+ * YAML {@link PropertySourceFactory} implementation, some source code is copied Spring Boot
+ * org.springframework.boot.env.YamlPropertySourceLoader , see {@link #createYaml()} and {@link #process()}
+ *
+ * @since 2.6.5
+ */
+public class YamlPropertySourceFactory extends YamlProcessor implements PropertySourceFactory {
```
Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/test/java/org/apache/dubbo/config/spring/beans/factory/config/YamlPropertySourceFactory.java]: It is no usage for other class. Why bring in this class ?

Comment (cvictory) [inline dubbo-config/dubbo-config-spring/src/test/java/org/apache/dubbo/config/spring/beans/factory/config/YamlPropertySourceFactory.java]: This is test for dubbo.yml but useless for dubbo-core.  It is just for fun?

**Comment by cvictory** [comment]
I did not review the unit test.

**Pull request merged**